manager/integration/tests/common.py

```diff
@@ -212,6 +233,122 @@ def get_compatibility_test_image(cli_v, cli_minv,
                                      data_v, data_minv)
 
 
+def generate_random_data(count):
+    return ''.join(random.choice(string.ascii_lowercase + string.digits)
+                   for _ in range(count))
+
+
+def volume_read(dev, start, count):
+    r_data = ""
+    fdev = open(dev, 'rb')
+    if fdev is not None:
+        fdev.seek(start)
+        r_data = fdev.read(count)
+        fdev.close()
+    return r_data
+
+
+def volume_write(dev, start, data):
+    w_length = 0
+    fdev = open(dev, 'rb+')
+    if fdev is not None:
+        fdev.seek(start)
+        fdev.write(data)
+        fdev.close()
+        w_length = len(data)
+    return w_length
+
+
+def volume_valid(dev):
+    return stat.S_ISBLK(os.stat(dev).st_mode)
+
+
+def parse_iscsi_endpoint(iscsi):
+    iscsi_endpoint = iscsi[8:]
+    return iscsi_endpoint.split('/')
+
+
+def get_iscsi_ip(iscsi):
+    iscsi_endpoint = parse_iscsi_endpoint(iscsi)
+    ip = iscsi_endpoint[0].split(':')
+    return ip[0]
+
+
+def get_iscsi_port(iscsi):
+    iscsi_endpoint = parse_iscsi_endpoint(iscsi)
+    ip = iscsi_endpoint[0].split(':')
+    return ip[1]
+
+
+def get_iscsi_target(iscsi):
+    iscsi_endpoint = parse_iscsi_endpoint(iscsi)
+    return iscsi_endpoint[1]
+
+
+def get_iscsi_lun(iscsi):
+    iscsi_endpoint = parse_iscsi_endpoint(iscsi)
+    return iscsi_endpoint[2]
+
+
+def exec_nsenter(cmd):
+    exec_cmd = "nsenter --mount=/host/proc/1/ns/mnt \
+               --net=/host/proc/1/ns/net bash -c \"" + cmd + "\""
+    fp = os.popen(exec_cmd)
+    ret = fp.read()
+    fp.close()
+    return ret
+
+
+def iscsi_login(iscsi_ep):
+    ip = get_iscsi_ip(iscsi_ep)
+    port = get_iscsi_port(iscsi_ep)
+    target = get_iscsi_target(iscsi_ep)
+    lun = get_iscsi_lun(iscsi_ep)
+    # discovery
+    cmd_discovery = "iscsiadm -m discovery -t st -p " + ip
+    exec_nsenter(cmd_discovery)
+    # login
+    cmd_login = "iscsiadm -m node -T " + target + " -p " + ip + " --login"
+    exec_nsenter(cmd_login)
+    blk_name = "ip-%s:%s-iscsi-%s-lun-%s" % (ip, port, target, lun)
+    wait_for_device_login(ISCSI_DEV_PATH, blk_name)
+    dev = os.path.realpath(ISCSI_DEV_PATH + "/" + blk_name)
+    return dev
+
+
+def iscsi_logout(iscsi_ep):
+    ip = get_iscsi_ip(iscsi_ep)
+    target = get_iscsi_target(iscsi_ep)
+    cmd_logout = "iscsiadm -m node -T " + target + " -p " + ip + " --logout"
+    exec_nsenter(cmd_logout)
+    cmd_rm_discovery = "iscsiadm -m discovery -p " + ip + " -o delete"
+    exec_nsenter(cmd_rm_discovery)
+
+
+def generate_random_pos(size, used):
+    start = 0
+    end = int(SIZE)
+    free = {}
+
+    if used is None:
+        position = random.randrange(start, end - size, 1)
+    else:
+        for u_start in sorted(used.keys()):
+            u_end = used[u_start]
+            if u_start > start:
+                free[start] = u_start
+            start = u_end
```

Comment:
This algorithm is correct, but it's unnecessarily complicated. We're writing a few bytes in a gigabyte-size disk. You can just generate a random number, calculate the start and the end and check if it collides with an existing position.

Reply:
I previously considered using the algorithm of check collision and gave up.
Now the algorithm of check collision can solve the problem quickly and easily. Please imagine that if you have allocated a lot of scattered positions, you will need to produce the position frequently, because the possibility of collision increases. But my algorithm can't solve the fragmentation problem.
I think we should add an assert when the position cannot be allocated.

Reply:
Do you mean hard to check collision using random location?

Currently the data we're going to write is a few hundred bytes. Total size of the volume is 30GB. The chance of collision is less than 1 in 30,000,000.

This algorithm works better if we have more constrain space to work with, though it still can't help with fragment if that need to be taken care of with constrain space.

The issue of this piece of code here is it's very easy to get it wrong and not easy to understand. Since there is an easy and straightforward alternative for our use case, I don't think it's necessary do it in this way.

And you're right. We need to make sure the position has been allocated.